src/prometheus/variant_counter.zig

```diff
@@ -0,0 +1,143 @@
+const std = @import("std");
+const prometheus = @import("lib.zig");
+
+const Atomic = std.atomic.Value;
+
+const Metric = prometheus.metric.Metric;
+const MetricType = prometheus.metric.MetricType;
+
+/// Separately count the occurrence of each variant within an enum or error set.
+///
+/// This is a composite metric, similar to a histogram, except the "buckets" are
+/// discrete, unordered, and identified by name, instead of representing numeric ranges.
+pub fn VariantCounter(comptime EnumOrError: type) type {
+    const indexer = VariantIndexer.init(EnumOrError);
+    const names = indexer.names();
+
+    return struct {
+        counts: [names.len]Atomic(u64) = .{Atomic(u64).init(0)} ** names.len,
+        metric: Metric = .{ .getResultFn = getResult },
+
+        const Self = @This();
+
+        pub const metric_type: MetricType = .variant_counter;
+        pub const Type = EnumOrError;
+
+        pub fn observe(self: *Self, err: EnumOrError) void {
+            _ = self.counts[indexer.index(err)].fetchAdd(1, .monotonic);
+        }
+
+        pub fn getResult(metric: *Metric, _: std.mem.Allocator) Metric.Error!Metric.Result {
+            const self: *Self = @fieldParentPtr("metric", metric);
+            return .{ .variant_counter = .{
+                .counts = &self.counts,
+                .names = &names,
+            } };
+        }
+    };
+}
+
+pub const VariantCounts = struct {
+    counts: []const Atomic(u64),
+    names: []const []const u8,
+};
+
+/// Assigns a continuous sequence of integers starting at 0 to the
+/// variants. For example, if you have an variant set with 10 variants,
+/// it will assign the numbers 0-9 to each of the variants.
+///
+/// Enums usually already have numbers with this property, but it's
+/// not guaranteed. This is particularly useful for variants, whose
+/// integers are likely not continuous.
+///
+/// You can provide an variant to determine its index, or provide an
+/// index to get the respective variant.
+///
+/// This struct must be initialized at comptime, but it can be used
+/// at any time.
+const VariantIndexer = struct {
+    EnumOrError: type,
```

Comment:
this originally confused me because i thought it was a struct which includes either enums or errors -- not one or the other -- maybe we rename to `Set` and then comptime ensure its either an enum or error set (similar to how we have logger) -- something like:
```
pub const Set = union(enum) {
    error_set: ErrorSet,
    enum: Enum,
```

Reply:
It could be a union but the whole idea of this field is to store the concrete type of the data. So the two variants of Set would need to have type `type`, and you'd need some way to extract the type out of it. Something like this:

```zig
pub const Set = union(enum) {
    error_set: type,
    enum: type,

    fn Type(self: Set) type {
        return switch(self) {
            .error_set => |e| e,
            .enum => |e| e,
        };
    }
}
```
and then most places where `self.EnumOrError` is used, you'll need to instead call it like `self.EnumOrError.Type()`

Reply:
mmm i see yeah thats not great - the way it is rn is ok